reload() was moved into importlib in Python 3

__reload()__ was moved into __importlib__ in Python 3 because it was being overuse and led issues that were to difficult-to-debug.

flake8 testing of https://github.com/1tayH/noisy on Python 3.6.3

$ __flake8 . --count --select=E901,E999,F821,F822,F823 --show-source --statistics__
```
./noisy.py:13:1: F821 undefined name 'reload'
reload(sys)
^
1     F821 undefined name 'reload'
1
```

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -53,4 +53,4 @@ workflows:
   build:
     jobs:
       - Python_2.7
-      - Python_3.6
+      #- Python_3.6

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -1,44 +1,59 @@
 version: 2
+
+my-steps: &steps
+  - checkout
+
+  # Download and cache dependencies
+  - restore_cache:
+      keys:
+      - v1-dependencies-{{ checksum "requirements.txt" }}
+      # fallback to using the latest cache if no exact match is found
+      - v1-dependencies-
+
+  - run:
+      name: install dependencies
+      command: |
+        mkdir -p ./venv
+        virtualenv ./venv
+        . venv/bin/activate
+        pip install -r requirements.txt
+
+  - save_cache:
+      paths:
+        - ./venv
+      key: v1-dependencies-{{ checksum "requirements.txt" }}
+
+  - run:
+      name: run script
+      # make sure the script runs, set a 60s timeout
+      command: |
+        . venv/bin/activate
+        python noisy.py --config config.json --timeout 60
+
+  - store_artifacts:
+      path: test-reports
+      destination: test-reports
+
 jobs:
+  Python_2.7:
+    build:
+      docker:
+        # specify the version you desire here
+        # use `-browsers` prefix for selenium tests, e.g. `3.6.1-browsers`
+        - image: circleci/python:2.7
+      working_directory: ~/repo
+      steps: *steps
+
+  Python_3.6:
+    build:
+      docker:
+        - image: circleci/python:3.6
+      working_directory: ~/repo
+      steps: *steps
+
+workflows:
+  version: 2
   build:
-    docker:
-      # specify the version you desire here
-      # use `-browsers` prefix for selenium tests, e.g. `3.6.1-browsers`
-      - image: circleci/python:2.7
-
-    working_directory: ~/repo
-
-    steps:
-      - checkout
-
-      # Download and cache dependencies
-      - restore_cache:
-          keys:
-          - v1-dependencies-{{ checksum "requirements.txt" }}
-          # fallback to using the latest cache if no exact match is found
-          - v1-dependencies-
-
-      - run:
-          name: install dependencies
-          command: |
-            mkdir -p ./venv
-            virtualenv ./venv
-            . venv/bin/activate
-            pip install -r requirements.txt
-
-      - save_cache:
-          paths:
-            - ./venv
-          key: v1-dependencies-{{ checksum "requirements.txt" }}
-
-      - run:
-          name: run script
-          # make sure the script runs, set a 60s timeout
-          command: |
-            . venv/bin/activate
-            python noisy.py --config config.json --timeout 60
-
-      - store_artifacts:
-          path: test-reports
-          destination: test-reports
-          
+    jobs:
+      - Python_2.7
+      - Python_3.6

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -53,4 +53,4 @@ workflows:
   build:
     jobs:
       - Python_2.7
-      #- Python_3.6
+      - Python_3.6

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -1,55 +1,52 @@
 version: 2
-
-my-steps: &steps
-  - checkout
-
-  # Download and cache dependencies
-  - restore_cache:
-      keys:
-      - v1-dependencies-{{ checksum "requirements.txt" }}
-      # fallback to using the latest cache if no exact match is found
-      - v1-dependencies-
-
-  - run:
-      name: install dependencies
-      command: |
-        mkdir -p ./venv
-        virtualenv ./venv
-        . venv/bin/activate
-        pip install -r requirements.txt
-
-  - save_cache:
-      paths:
-        - ./venv
-      key: v1-dependencies-{{ checksum "requirements.txt" }}
-
-  - run:
-      name: run script
-      # make sure the script runs, set a 60s timeout
-      command: |
-        . venv/bin/activate
-        python noisy.py --config config.json --timeout 60
-
-  - store_artifacts:
-      path: test-reports
-      destination: test-reports
-
 jobs:
   Python_2.7:
-    build:
-      docker:
-        # specify the version you desire here
-        # use `-browsers` prefix for selenium tests, e.g. `3.6.1-browsers`
-        - image: circleci/python:2.7
-      working_directory: ~/repo
-      steps: *steps
+    docker:
+      # specify the version you desire here
+      # use `-browsers` prefix for selenium tests, e.g. `3.6.1-browsers`
+      - image: circleci/python:2.7
+
+    working_directory: ~/repo
+
+    steps: &steps
+      - checkout
+
+      # Download and cache dependencies
+      - restore_cache:
+          keys:
+          - v1-dependencies-{{ checksum "requirements.txt" }}
+          # fallback to using the latest cache if no exact match is found
+          - v1-dependencies-
+
+      - run:
+          name: install dependencies
+          command: |
+            mkdir -p ./venv
+            virtualenv ./venv
+            . venv/bin/activate
+            pip install -r requirements.txt
+
+      - save_cache:
+          paths:
+            - ./venv
+          key: v1-dependencies-{{ checksum "requirements.txt" }}
+
+      - run:
+          name: run script
+          # make sure the script runs, set a 60s timeout
+          command: |
+            . venv/bin/activate
+            python noisy.py --config config.json --timeout 60
+
+      - store_artifacts:
+          path: test-reports
+          destination: test-reports
 
   Python_3.6:
-    build:
-      docker:
-        - image: circleci/python:3.6
-      working_directory: ~/repo
-      steps: *steps
+    docker:
+      - image: circleci/python:3.6
+    working_directory: ~/repo
+    steps: *steps
 
 workflows:
   version: 2

--- a/README.md
+++ b/README.md
@@ -4,7 +4,7 @@
 
 A simple python script that generates random HTTP/DNS traffic noise in the background while you go about your regular web browsing, to make your web traffic data less valuable for selling and for extra obscurity.
 
-Tested on MacOS High Sierra, Ubuntu 16.04 and Raspbian Stretch
+Tested on MacOS High Sierra, Ubuntu 16.04 and Raspbian Stretch and is compatable with both Python 2.7 and 3.6
 
 ## Getting Started
 

--- a/noisy.py
+++ b/noisy.py
@@ -10,17 +10,15 @@ import time
 import requests
 
 try:                 # Python 2
-    reload
-except NameError:    # Python 3
-    from importlib import reload
-
-try:                 # Python 2
     from urllib.parse import urljoin, urlparse
 except ImportError:  # Python 3
     from urlparse import urljoin, urlparse
 
-reload(sys)
-sys.setdefaultencoding('latin-1')
+try:                 # Python 2
+    reload(sys)
+    sys.setdefaultencoding('latin-1')
+excpet NameError:    # Python 3
+    pass
 
 
 class Crawler(object):

--- a/noisy.py
+++ b/noisy.py
@@ -1,14 +1,23 @@
-import requests
-import re
-import time
-import random
-import logging
 import argparse
-import json
-import urlparse
-import sys
 import datetime
+import json
+import logging
+import random
+import re
+import sys
+import time
 
+import requests
+
+try:                 # Python 2
+    reload
+except NameError:    # Python 3
+    from importlib import reload
+
+try:                 # Python 2
+    from urllib.parse import urljoin, urlparse
+except ImportError:  # Python 3
+    from urlparse import urljoin, urlparse
 
 reload(sys)
 sys.setdefaultencoding('latin-1')
@@ -52,8 +61,8 @@ class Crawler(object):
         :param root_url: the URL the DOM was loaded from
         :return: absolute link
         """
-        parsed_url = urlparse.urlparse(link)
-        parsed_root_url = urlparse.urlparse(root_url)
+        parsed_url = urlparse(link)
+        parsed_root_url = urlparse(root_url)
 
         # '//' means keep the current protocol used to access this URL
         if link.startswith("//"):
@@ -61,7 +70,7 @@ class Crawler(object):
 
         # possibly a relative path
         if not parsed_url.scheme:
-            return urlparse.urljoin(root_url, link)
+            return urljoin(root_url, link)
 
         return link
 

--- a/noisy.py
+++ b/noisy.py
@@ -140,7 +140,7 @@ class Crawler(object):
         :param depth: our current link depth
         """
         is_depth_reached = depth >= self._config['max_depth']
-        if not len(self._links) or is_depth_reached:
+        if not len(list(self._links)) or is_depth_reached:
             logging.debug("Hit a dead end, moving to the next root URL")
             # escape from the recursion, we don't have links to continue or we have reached the max depth
             return

--- a/noisy.py
+++ b/noisy.py
@@ -228,7 +228,7 @@ class Crawler(object):
                 try:
                     body = self._request(url).content
                     self._links = self._extract_urls(body, url)
-                    logging.debug("found {} links".format(len(self._links)))
+                    logging.debug("found {} links".format(len(list(self._links))))
                     self._browse_from_links()
 
                 except requests.exceptions.RequestException:

--- a/noisy.py
+++ b/noisy.py
@@ -118,7 +118,7 @@ class Crawler(object):
         urls = re.findall(pattern, str(body))
 
         normalize_urls = [self._normalize_link(url, root_url) for url in urls]
-        filtered_urls = filter(self._should_accept_url, normalize_urls)
+        filtered_urls = list(filter(self._should_accept_url, normalize_urls))
 
         return filtered_urls
 
@@ -140,7 +140,7 @@ class Crawler(object):
         :param depth: our current link depth
         """
         is_depth_reached = depth >= self._config['max_depth']
-        if not len(list(self._links)) or is_depth_reached:
+        if not len(self._links) or is_depth_reached:
             logging.debug("Hit a dead end, moving to the next root URL")
             # escape from the recursion, we don't have links to continue or we have reached the max depth
             return
@@ -228,7 +228,7 @@ class Crawler(object):
                 try:
                     body = self._request(url).content
                     self._links = self._extract_urls(body, url)
-                    logging.debug("found {} links".format(len(list(self._links))))
+                    logging.debug("found {} links".format(len(self._links)))
                     self._browse_from_links()
 
                 except requests.exceptions.RequestException:

--- a/noisy.py
+++ b/noisy.py
@@ -17,7 +17,7 @@ except ImportError:  # Python 3
 try:                 # Python 2
     reload(sys)
     sys.setdefaultencoding('latin-1')
-excpet NameError:    # Python 3
+except NameError:    # Python 3
     pass
 
 

